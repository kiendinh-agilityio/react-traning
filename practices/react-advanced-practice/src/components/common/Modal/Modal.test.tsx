import { render, fireEvent } from '@testing-library/react';
import Modal from '.';

describe('Modal Component', () => {
  const mockOnClose = jest.fn();

  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('renders correctly and matches snapshot', () => {
    const { asFragment } = render(
      <Modal onClose={mockOnClose}>
        <div>Modal Content</div>
      </Modal>,
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('does not call onClose when clicking inside modal', () => {
    const { asFragment, container } = render(
      <Modal onClose={mockOnClose}>
        <div>Modal Content</div>
      </Modal>,
    );
    const modalContent = container.querySelector('.bg-white');
    if (modalContent) {
      fireEvent.click(modalContent);
    }
    expect(mockOnClose).not.toHaveBeenCalled();
    expect(asFragment()).toMatchSnapshot();
  });

  it('calls onClose when clicking on overlay', () => {
    const { asFragment, container } = render(
      <Modal onClose={mockOnClose}>
        <div>Modal Content</div>
      </Modal>,
    );
    const overlay = container.querySelector('.bg-overlay');
    if (overlay) {
      fireEvent.click(overlay);
    }
    expect(mockOnClose).toHaveBeenCalledTimes(1);
    expect(asFragment()).toMatchSnapshot();
  });

  it('handles stopPropagation correctly', () => {
    const stopPropagationMock = jest.fn();
    const { asFragment, container } = render(
      <Modal onClose={mockOnClose}>
        <div className="modal-content" onClick={(e) => stopPropagationMock(e)}>
          Modal Content
        </div>
      </Modal>,
    );
    const modalContent = container.querySelector('.modal-content');
    if (modalContent) {
      fireEvent.click(modalContent);
    }
    expect(stopPropagationMock).toHaveBeenCalled();
    expect(asFragment()).toMatchSnapshot();
  });
});
