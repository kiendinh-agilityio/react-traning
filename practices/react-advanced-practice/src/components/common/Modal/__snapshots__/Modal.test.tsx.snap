// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal Component calls onClose when clicking on overlay 1`] = `
<DocumentFragment>
  <div
    class="rt-Flex rt-r-ai-center rt-r-jc-center bg-overlay fixed top-0 left-0 right-0 bottom-0"
  >
    <div
      class="rt-Flex rt-r-fd-column gap-8 rounded-2xl bg-white shadow z-10 py-8 px-10 undefined"
    >
      <div>
        Modal Content
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Modal Component does not call onClose when clicking inside modal 1`] = `
<DocumentFragment>
  <div
    class="rt-Flex rt-r-ai-center rt-r-jc-center bg-overlay fixed top-0 left-0 right-0 bottom-0"
  >
    <div
      class="rt-Flex rt-r-fd-column gap-8 rounded-2xl bg-white shadow z-10 py-8 px-10 undefined"
    >
      <div>
        Modal Content
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Modal Component handles stopPropagation correctly 1`] = `
<DocumentFragment>
  <div
    class="rt-Flex rt-r-ai-center rt-r-jc-center bg-overlay fixed top-0 left-0 right-0 bottom-0"
  >
    <div
      class="rt-Flex rt-r-fd-column gap-8 rounded-2xl bg-white shadow z-10 py-8 px-10 undefined"
    >
      <div
        class="modal-content"
      >
        Modal Content
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Modal Component renders correctly and matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="rt-Flex rt-r-ai-center rt-r-jc-center bg-overlay fixed top-0 left-0 right-0 bottom-0"
  >
    <div
      class="rt-Flex rt-r-fd-column gap-8 rounded-2xl bg-white shadow z-10 py-8 px-10 undefined"
    >
      <div>
        Modal Content
      </div>
    </div>
  </div>
</DocumentFragment>
`;
